safehttp/plugins/collector/collector.go

```diff
@@ -0,0 +1,236 @@
+	// occurred.
+	ColumnNumber uint
+}
+
+// HandlerBuilder creates a safehttp.Handler based on the report handlers that
+// it's created with.
+type HandlerBuilder struct {
+	// Handler handles all generic reports received with the Content-Type application/reports+json.
+	Handler func(Report)
+	// CSPHandler handles all CSP reports received with the Content-Type application/csp-report.
+	CSPHandler func(CSPReport)
+}
+
+// Build builds a safehttp.Handler which calls the given Handler or CSPHandler when
+// a violation report is received. Make sure to register the handler to receive POST
+// requests. If the handler recieves anything other than POST requests it will
+// respond with a 405 Method Not Allowed.
+func (hb HandlerBuilder) Build() safehttp.Handler {
+	return safehttp.HandlerFunc(func(w *safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
+		if r.Method() != safehttp.MethodPost {
+			return w.ClientError(safehttp.StatusMethodNotAllowed)
+		}
+
+		b, err := ioutil.ReadAll(r.Body())
+		if err != nil {
+			return w.ClientError(safehttp.StatusBadRequest)
+		}
+
+		ct := r.Header.Get("Content-Type")
+		if ct == "application/csp-report" && hb.CSPHandler != nil {
+			return handleDeprecatedCSPReports(hb.CSPHandler, w, b)
+		} else if ct == "application/reports+json" && hb.Handler != nil {
+			return handleReport(hb.Handler, w, b)
+		}
+
+		return w.ClientError(safehttp.StatusUnsupportedMediaType)
+	})
+}
+
+func stringOrEmpty(x interface{}) string {
+	s, ok := x.(string)
+	if !ok {
+		return ""
+	}
+	return s
+}
+
+func uintOrZero(x interface{}) uint {
+	// x must be a float64 that will be converted to a uint since this is how json stores
+	// its numbers.
+	f, ok := x.(float64)
+	if !ok || f < 0 {
+		return 0
+	}
+	return uint(f)
+}
+
+func handleDeprecatedCSPReports(h func(CSPReport), w *safehttp.ResponseWriter, b []byte) safehttp.Result {
+	// In CSP2 it is clearly stated that a report has a single key 'csp-report'
+	// which holds the report object. Like this:
+	// {
+	//   "csp-report": {
+	//     // report goes here
+	//   }
+	// }
+	// Source: https://www.w3.org/TR/CSP2/#violation-reports
+	//
+	// But in the CSP3 spec this 'csp-report' key is never mentioned. So the report
+	// would look like this:
+	// {
+	//   // report goes here
+	// }
+	// Source: https://w3c.github.io/webappsec-csp/#deprecated-serialize-violation
+	//
+	// Because of this we have to support both. :/
+	m := make(map[string]interface{})
+	if err := json.Unmarshal(b, &m); err != nil {
+		return w.ClientError(safehttp.StatusBadRequest)
+	}
+
+	if report, exists := m["csp-report"]; exists {
+		var ok bool
+		m, ok = report.(map[string]interface{})
+		if !ok {
+			return w.ClientError(safehttp.StatusBadRequest)
+		}
+	}
+
+	ln := uintOrZero(m["lineno"])
+	if ln == 0 {
+		ln = uintOrZero(m["line-number"])
+	}
+	cn := uintOrZero(m["colno"])
+	if cn == 0 {
+		cn = uintOrZero(m["column-number"])
+	}
```

Comment:
I would suggest to not use this.

I would unmarshal on a local type that has all the fields, including both ways to say the same thing, and when you construct `CSPReport` you just copy the fields over. This will take care of alternative ways and correct type casting.

e.g.
```go
// inside the func
tmp := struct {
LineNumber int `json:"line-number"`
Lineno int
}{}
// unmarshal
r := CSPReport { LineNumber: tmp.LineNumber+tmp.Lineno }
```